File: Java/0151-0200/152. Maximum Product Subarray.java
// https://leetcode.com/problems/maximum-product-subarray/description/

class Solution {
    public int maxProduct(int[] nums) {
        int result = nums[0] , maxProduct = nums[0], minProduct = nums[0];

        for (int index = 1;index < nums.length; index++) {

            int temp = maxProduct;
            maxProduct = Math.max(nums[index], Math.max(nums[index] * maxProduct, nums[index] * minProduct));
            minProduct = Math.min(nums[index], Math.min(nums[index] * minProduct, nums[index] * temp));

            result = Math.max(result, maxProduct); 
        }
        return result;
    }
}

File: Python/0151-0200/152. Maximum Product Subarray.py
# https://leetcode.com/problems/maximum-product-subarray/description/

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        result = maxProduct = minProduct = nums[0]

        for index in range(1, len(nums)):
            temp = maxProduct
            maxProduct = max(nums[index], max(nums[index] * maxProduct, nums[index] * minProduct))
            minProduct = min(nums[index], min(nums[index] * minProduct, nums[index] * temp))

            # update result
            result = max(result, maxProduct)

        return result
